orchagent/routeorch.cpp

```diff
@@ -1227,14 +1227,46 @@ bool RouteOrch::addRoute(RouteBulkContext& ctx, const NextHopGroupKey &nextHops)
     sai_object_id_t& vrf_id = ctx.vrf_id;
     IpPrefix& ipPrefix = ctx.ip_prefix;
 
-    if (m_fgNhgOrch->fgNhgPrefixes.find(ipPrefix) != m_fgNhgOrch->fgNhgPrefixes.end()
-            && vrf_id == gVirtualRouterId)
+    if (m_fgNhgOrch->isRouteFineGrainedECMP(vrf_id, ipPrefix, nextHops))
     {
-        /* Only support the default vrf for Fine Grained ECMP */
-        SWSS_LOG_INFO("Reroute %s:%s to fgNhgOrch", ipPrefix.to_string().c_str(), 
-                nextHops.to_string().c_str());
+        /* The route needs Fine Grained ECMP, first check if routeorch has a route
+         * matching the prefix. If routeorch does then we need to delete the route
+         * so that it can be added for fine grained ecmp after that. */ 
+        auto route_table = m_syncdRoutes.find(vrf_id);
+        if (route_table != m_syncdRoutes.end())
+        {
+            auto route_entry = route_table->second.find(ipPrefix);
+            if (route_entry != route_table->second.end()) 
+            {
+                /* Case where we transition route from Regular ECMP to Fine Grained ECMP */
+                NextHopGroupKey nhgk = route_entry->second;
+                removeRoute(ctx);
```

Comment:
>removeRoute [](start = 16, length = 11)

I am worried about the complex design.
We should not call `removeRoute` inside `addRoute`
We should not manipulate `gRouteBulker` inside `addRoute`.

Reply:
This is to handle the case where the route needs Fine Grained ECMP based on the nexthops of the route. The only other way I see to handle this(not do removeRoute/bulker calls) is to add checks in fgnhgorch to see if a route exists in routeorch already, and if so use sai apis to modify the nexthop of the route(set_entry_attr) to fine grained ecmp, but this would mean that fgnhgorch will need to call into routeorch to delete the routes from routeorch internal data structure and remove the next-hop group if the ref count drops to 0. IMO the full routeorch removeRoute, as implemented just now, seems cleaner to me and better from a maintainability standpoint, though I am open to the mentioned way/ other suggestions.

Reply:
Made changes for compatibility and to address the concerns